### ISSUE TITLE
Mesh specification

### ISSUE THREAD

**Issue by glyg** [ome/ngff#33]
As discussed in the feb. 2021 ngff [community call](https://forum.image.sc/t/next-call-on-next-gen-bioimaging-data-tools-feb-23/48386), and following this [image.sc thread](https://forum.image.sc/t/mesh-data-in-ome-zarr/)

The idea is to follow PLY specification to store meshes in ome-zarr. A ply file is organised in:
- a header (translated in the .attrs of the zarr group)
- the points (a zarr.Array)
- for each element (e.g. faces) the list of vertices. Those are group by number of sides, with one zarr.Array per polygon.

There is a draft implementation here: https://github.com/centuri-engineering/ply-zarr

Some questions: 

* I don't know how to deal with binary formats 
* PLY allows for datatype specification at the column level, how to deal with that in zarr?

**Comment by joshmoore** [comment]
cc: @jfkotw @kephale who were also interested during the meeting. I defer on whether or not this issue covers _all_ of "vector-based".

**Comment by glyg** [comment]
I'd argue for GEOjson for ROIs and points and such & keep meshes in their niche

**Comment by kephale** [comment]
PLY also allows for mixing of faces with different number of vertices (e.g. triangles and quads).

**Comment by joshmoore** [comment]
@glyg, so this block from ply-zarr is the critical bit for discussion?

```
ply_header = {
    "format": "ascii 1.0",
    "comments": [f"created by ply_zarr v0.0.1, {datetime.now().isoformat()}",],
    "elements": {
        "vertex": {
            "size": 47,
            "properties": [
                ("double", "x"),
                ("double", "y"),
                ("double", "z")
            ]
        },
        "face": {
            "size": 105,
            "properties": [
                ("list", "uint8", "int32", "vertex_indices"),
            ]
        }
    }
}
```

**Comment by glyg** [comment]
Yes, this mirrors the specification for the PLY header, then it seems natural to store the faces in separate arrays according to their number of sides.

**Comment by glyg** [comment]
see a more concrete example of mixing meshes, images and labels [here](https://github.com/centuri-engineering/ply-zarr/blob/main/doc/Image_and_surface.ipynb)

I assume the xarray compatibility also applies here, I'll look into that next.

**Comment by jburel** [comment]
cc @normanrz

**Comment by normanrz** [comment]
We recently implemented the mesh format from Neuroglancer: https://github.com/google/neuroglancer/blob/master/src/neuroglancer/datasource/precomputed/meshes.md

It's been great for our purposes:
- Supports billions of meshes
- Uses draco encoding, which is very storage-efficient
- Supports chunked mesh loading, which is important for very large objects
- Supports multi level-of-detail for progressive loading

I think that format would be a great candidate to be adopted by OME-NGFF.

**Comment by glyg** [comment]
@normanrz thanks for the input, those features indeed sound great (esp. multi-res!). 

If I understand correctly though, only triangular meshes are supported? The other consumer / producer of meshes is the modeling community (i.e. physical biology), who would need more generic meshes, for example with polygonal (>3) 2D cells, polyhedral 3D cells,  or even quadratic tetrahedra. 

Would draco  be able to handle that kind of data? 
How would a zarr implementation work? Is it enough to "just" put the draco encoded in the store and add a dependency to be able to read / write them?

Also, maybe storing generic FEM meshes is out of scope for ome-ngff and triangles are enough.

**Comment by normanrz** [comment]
> If I understand correctly though, only triangular meshes are supported? The other consumer / producer of meshes is the modeling community (i.e. physical biology), who would need more generic meshes, for example with polygonal (>3) 2D cells, polyhedral 3D cells, or even quadratic tetrahedra.

Yes, I think draco only supports triangular meshes (and point clouds). We could look into allowing other encodings in addition to draco.

 > How would a zarr implementation work? Is it enough to "just" put the draco encoded in the store and add a dependency to be able to read / write them?

That is a good question that we haven't fully figured out yet. We currently store all the data in a single binary file. The file consists of a) a directory structure (hash map) to locate the meshfile for a specified id within b) a long blob of mesh data. In b) each meshfile has a binary metadata header that describes the available chunks and level-of-details. 
One implementation on top of zarr would be to store each meshfile as on chunk (e.g. in a 2D uint8 array). This would create a lot of chunk files and might create some issues, because the chunks will have different byte lengths.

**Comment by EricMoerthVis** [comment]
I would like to get involved in the discussion.

I think it would be great to have a format similar to the Neuroglancer format in OME. Now the 3D data generation is getting more and more popular in the Spatial Biology field and segmentations are a big part of it. Having the possibility to, beside storing the volumetric (point cloud) data in OME Zarr it would be greally great to have the same possibility to do that for meshes.

I am wondering if there would be the possibility to exchange about the format and specifications in a meeting or such?

**Comment by joshmoore** [comment]
> I would like to get involved in the discussion.

Consider yourself involved! 🙂 

> I think it would be great to have a format similar to the Neuroglancer format in OME.

Modulo https://xkcd.com/927/ of course. This is certainly something that I've heard several times recently as well, but it will certainly take one or more champions for it to happen. Also cc @jbms for how he weighs the changes as well as the pros & cons.

> I am wondering if there would be the possibility to exchange about the format and specifications in a meeting or such?

Most of the recent meetings have been around the [challenge](https://forum.image.sc/t/ome2024-ngff-challenge/97363/1) which is pushing forward Zarr v3 support (i.e., [RFC-2](https://ngff.openmicroscopy.org/rfc/2/)).  It's certainly time for a next general community meeting, or alternatively, a smaller group could start *[socializing](https://ngff.openmicroscopy.org/rfc/1/index.html#stakeholders)* the idea in preparation for a RFC.

**Comment by kephale** [comment]
I'm still here watching this thread, and would be happy to help get a small group discussing what the best options are for this!

**Comment by EricMoerthVis** [comment]
I see!

I think there are similarities and differences between storing volumetric (point cloud data) and the meshes. One main similarity as introduced by the standard Neuroglancer uses is:

Multi-Resolution support for meshes! This is really crucial for the vast amount of meshes we are gonna store and load again

I think the main difference is that meshes don't adhere to such a nice Grid Structure as the point Clouds. So I am wondering how we can store them in their multi resolutions but still know where they are located in XYZ so we can efficiently load them when needed.

So there might be more MetaData to know the Bounding Box, Centroid or other measures to know if a Mesh is visible in a certain location so we can define if it should be loaded by the client or not.

Would really like to see a first Mesh Support (maybe based on the NeuroGlancer format supporting Draco) soon in Zarr

**Comment by d-v-b** [comment]
> Would really like to see a first Mesh Support (maybe based on the NeuroGlancer format supporting Draco) soon in Zarr

What would meshes look like in the Zarr data model? Zarr v3 doesn't have support yet for variable length types, so at a minimum we would need to add that, and even then I'm not sure how meshes, expressed as variable-length collections of geometrical objects, would be stored in an N-dimensional array. What would the array indices mean? I suspect people would fall back to 1D arrays, with maybe a second array for storing a spatial index? It could work, but it's not a great fit for Zarr IMO.

On the other hand, the neuroglancer multiresolution mesh format seems perfectly fine on its own, outside of Zarr. So maybe just refining or generalizing that format as needed would be simpler than forcing it into Zarr.

**Comment by normanrz** [comment]
I agree that the mesh format doesn't need to live in Zarr arrays. We could (mis)use uint8 arrays, to store the bytes, but I don't know what value that would bring in comparison to just storing the blob alongside the Zarr arrays in the hierarchy.
In general, I don't think that all pieces of OME-Zarr need to be Zarr.

**Comment by EricMoerthVis** [comment]
So the idea would be to adopt the NeuroGlancer Format (https://github.com/google/neuroglancer/blob/master/src/datasource/precomputed/meshes.md#multi-resolution-mesh-format) and integrate it into OME-Zarr?

**Comment by normanrz** [comment]
> So the idea would be to adopt the NeuroGlancer Format (https://github.com/google/neuroglancer/blob/master/src/datasource/precomputed/meshes.md#multi-resolution-mesh-format) and integrate it into OME-Zarr?

I think that would be a good way forward. There are a few details in terms of metadata and file layout that need to be figured out. Would be great to hear @jbms feedback on this.

**Comment by joshmoore** [comment]
A quick heads up that I heard from Jeremy today on a separate matter: he's been on leave. I very much assume when he's caught back up he'll chime in.

**Comment by EricMoerthVis** [comment]
I just want to get this discussion running again. What would be potential next steps?

**Comment by normanrz** [comment]
I think a meeting to sketch out an RFC would be a good next step. There should be an accompanying post on image.sc to announce that meeting.

**Comment by d-v-b** [comment]
@normanrz I'm not sure how crystallized the schedule is for the upcoming OME-NGFF workflows hackathon, but maybe carving out some (EST-timezone-friendly) slots would be convenient?

**Comment by EricMoerthVis** [comment]
That sounds like a good plan to discuss in that timeframe!

**Comment by jbms** [comment]
Sorry, was on paternity leave until today.

As others have also stated, while meshes can be potentially thought of as collections of arrays of vertex properties and face properties, I think trying to represent them as zarr arrays directly would add a lot of complexity and not provide significant advantages, given how meshes are actually used in practice.

There is certainly a lot of room for improvement in the Neuroglancer precomputed multiscale mesh format (and the related annotation format) but I think if the existing format serves a decent number of use cases then it may be wise to standardize it as-is initially, and then once there is greater usage experience work on a revised format.

**Comment by EricMoerthVis** [comment]
No worries!

Yes! I think this sounds like a really good plan! I think there is also a great need for more standardized creation and retrieving pipelines for the format. So I like your suggestion of first taking it up as-is and gradually improving it over time.

**Comment by Eric-Spooner** [comment]
Just wanting to give this issue a new bump, could we potentially find a meeting time in the week of the 16th of December to discuss next steps here?

**Comment by EricMoerthVis** [comment]
Just wanting to give this issue a new bump, could we potentially find a meeting time in the week of the 16th of December to discuss next steps here?

**Comment by joshmoore** [comment]
I'd be on board.

**Comment by normanrz** [comment]
Yes, I also have some time.

**Comment by kephale** [comment]
LOL ok fine I'll start the meeting scheduler :P

https://www.when2meet.com/?27988984-tS94j

Let's select the time on Sunday.

**Comment by kephale** [comment]
Bumping @joshmoore @normanrz @EricMoerthVis @d-v-b @jbms

**Comment by joshmoore** [comment]
Yay! Other community co-organizers very appreciated!! 🙌🏽

**Comment by kephale** [comment]
Let's do both:
- 10a Dec. 18 EST
- 1p Dec 19 EST

**Comment by EricMoerthVis** [comment]
Both dates and times sound good! Thank you for organizing!

**Comment by joshmoore** [comment]
> I don't have folks email addresses (except Josh's Dundee account and Jeremy).

Made use of the `@ngff` "mailing list" on https://forum.image.sc/t/late-announcement-technical-ngff-call-on-meshes/106304

**Comment by unidesigner** [comment]
I'd like to second @d-v-b @jbms @normanrz from previous comments about not using zarr for mesh representation and supporting e.g. Draco as "the standard". What is not immediately clear to me when looking at e.g. DracoPy though is the degree it supports storing arbitrary numeric attributes on the vertices and faces in an efficient way, [this issue](https://github.com/seung-lab/DracoPy/issues/22) may suggest perhaps not. Having a separate file for attributes, e.g. based on Parquet, to efficiently store/retrieve attribute values matched to indices may be useful. But it'd need to play nicely e.g. with the shaders in Neuroglancer to visualize the values in 3d.

And is a form of sharding supported for storing multiple multi-resolution meshes?

I'll try to join the call tomorrow.

**Comment by EricMoerthVis** [comment]
Yes, I think the NeuroGlancer Precomputed Multi-Res Sharded Format is the way to go. It would benefit from better tool support and better documentation. We could think about using the IDs that index the meshes to link additional information to the meshes. I think NeuroGlancer has some sort of additional attribute display for segmentations (https://github.com/google/neuroglancer/issues/430).

**Comment by unidesigner** [comment]
The additional attributes are on an object like, i.e. for segmentation objects and their associated meshes. What I meant are attributes per mesh vertex/face. Good to know that draco can support that. I'm sure there will be many use cases one can think of, and having tools to visualize is crucial. [Vedo](https://vedo.embl.es/#gallery) is nice Python library for "small" scale use cases. 

It may already exist, but it would be useful to have an easy way (e.g. simple Python function) to go from a set of mesh files to NG Precomputed Multi-Res Sharded Format, perhaps with the option for downscaling/simplification levels and specification of some affine transformation. I know [igneous](https://github.com/seung-lab/igneous?tab=readme-ov-file#in-memory-queue-simple-execution) can write it, but it's specifically designed for large-scale data use cases.

Also it may be useful to discuss, in the context of OME-Zarr, how transformations of meshes (i.e. the transformation-spec) could be used with or without associated image arrays.

**Comment by kephale** [comment]
@jbms TL;DR for you from [the meeting notes]([hackmd.io/@kisharrington/SJBT58eHke](https://hackmd.io/@kisharrington/SJBT58eHke)): 
How do you feel about NGFF using NG precomputed as the starting point for meshes in NGFF? 

One thing that might inform your response: at the end of the meeting notes, @bogovicj pointed out some specific technical questions that would need to be addressed (where/how transforms are stored relative to current NGFF approach, and axis names/order).

**Comment by EricMoerthVis** [comment]
Unfortunately I cannot join the call today due to sickness. Will follow the outcomes via the notes and summaries. Thank you for the discussion and effort to get meshes supported 🙏

**Comment by kephale** [comment]
We had 2 great meetings. Notes [here.](https://hackmd.io/P40be9p4SdGuPkhKKNnp7A)

Some notes:
- We need a RFC for meshes now.
- We all agree that NG precomputed should be the starting point. There may be some consideration for considering simple representations (e.g. `ply`).
- We need to relate the metadata from NG with NGFF (e.g. transforms and axes).

Some considerations:
- NG doesn't support arbitrary per-vertex/per-face attributes (but could, Draco supports multiple attributes).
- Not many tools available for NG meshes at the moment (neuroglance, webknossos, igneous, ?)
- Only supports triangular meshes
- How does this relate to NGFF collections?

Some goals of the first effort:
- Aim to address the `nm` unit assumption in NG and axes assumptions.
- Make a URL with a zarr that contains a mesh, show that both NG and webknossos can render the data with the same relationships including the mesh.
- Improve the utility/tool landscape for NG meshes (see notes for links to multiple writers)
- Draft spec/RFC

**Comment by kephale** [comment]
Okey doke, I tossed together a first draft of the RFC: https://github.com/kephale/ngff/blob/mesh/rfc/7/index.md

I want to start playing around with an implementation before getting too carried away with the text.

**Comment by jbms** [comment]
Thanks for putting that together.  I have a few comments: 

- mesh.json vs zarr.json: since the json seems to be intended to be a valid zarr group, I think it would make more sense to name it zarr.json.  Additionally, though, we might want the zarr node type to be something other than group, e.g. some new mesh node type, since it is not in fact intended to contain other zarr nodes as children. 

- the directory layout shown differs from the current format.  Is that intentional?

**Comment by kephale** [comment]
Aha, good point about the naming. I swapped the json and node_type name to zarr.json and `node_type: "mesh"`.

@jbms the directory structure is written naively tbh. I'm happy to adjust back, but I was going to update after testing implementation.

Incidentally, I tried [zmesh](https://github.com/seung-lab/zmesh) outputs to ng precomputed but it doesn't seem to match the spec (afaict it just outputs the fragments): https://github.com/kephale/atrium.kyleharrington.com/blob/main/meshes/zmesh-ng-precomputed/0.0.1.py

I'll play other writers soon.

**Comment by jbms** [comment]
Yes, zmesh only seems to support the older single-resolution neuroglancer precomputed mesh format and indeed seems to also skip writing the manifests.

**Comment by normanrz** [comment]
Thanks @kephale for writing this up!
There is some ongoing work on a collections proposal for OME-Zarr. That will likely replace the special `labels` group for specifying image-segmentation relationships. We should sync with that.
I am not sure I like `node_type: mesh`. Most zarr libraries will fail when parsing this, which might make it harder to implement this.
I am on the fence whether the mesh should be a Zarr node at all. It could be its own artifact type that happens to be stored within a Zarr hierarchy. From the upcoming collections spec, it would be addressable by path.

**Comment by kephale** [comment]
In which case, we might switch the metadata file back to `mesh.json` or even keep it as `info` like the original neuroglancer spec.

@normanrz we'll also be able to piggy back on the coordinate transforms support from Collections as well, right?

**Comment by jbms** [comment]
I suppose there are a lot of options for the metadata.

In general people may wish to store non-zarr data within a zarr hierarchy so a general solution to that may be useful.

Several possible solutions come to mind:

- Specifying in the zarr spec that for any unknown node type, the attributes can still be found under the `attributes` key.
- Creating a new node type of `external` that indicates non-zarr data can be found in the directory, but attributes can still be found under the `attributes` key.

Alternatively, now that implicit groups are disallowed, we could say that any directory that lacks a zarr.json is assumed to be unknown "external" data.  However, if creating a new group gets interrupted it would appear as an "external" node which may be problematic.

**Comment by joshmoore** [comment]
> * Creating a new node type of `external` that indicates non-zarr data can be found in the directory, but attributes can still be found under the `attributes` key.

*nods* or "folder" / "other" etc. might all specify this is a directory in the hierarchy that shouldn't be parsed but it still has metadata in it, but **(sidenote)** that discussion we should really have with the wider Zarr community.

**Comment by normanrz** [comment]
I was wondering if you had thoughts on using sqlite as a container for the mesh manifest and fragment files instead of the `neuroglancer_uint64_sharded_v1` hash map? The advantage would be that we could use a standard format with existing tools to create and read the mesh files. 

There are some examples where read was data from a remotely stored sqlite database [1](https://news.ycombinator.com/item?id=32828799) [2](https://github.com/s3lite/s3lite) [3](https://github.com/litements/s3sqlite). An alternative might be [DuckDB](https://duckdb.org/docs/guides/network_cloud_storage/duckdb_over_https_or_s3.html).

**Comment by jbms** [comment]
I agree that the non-standard neuroglancer precomputed sharded format is not ideal.  However, for scalability what is needed here is a "cloud-native" key-value store with the following properties:

- packs more than one key/value pair in a single file, to avoid an excessive number of small files
- supports efficiently reading arbitrary keys in parallel, from high-latency storage, at high throughput, without requiring a large amount of metadata to be loaded up front
- supports efficient parallel writing at high throughput from multiple machines

There were some of the main design criteria for OCDBT, which I think would be a good candidate for use instead of the precomputed sharded format.

There are a few problems with sqlite:
- due to the nature of the sqlite format, it is basically impossible to write in parallel from multiple machines, and therefore it would need to be sharded for efficient writing, which would be an unfortunate complication
- in principle, it would be possible to read separate keys efficiently in parallel.  However, the only implementation of sqlite of which I'm aware, the C implementation, is inherently synchronous, and always requests just a single page at a time, and does not request another page until the previously-requested page has been received.  Page size is configurable at database creation time from 512 to 65536 bytes, but regardless, for high-latency storage, reading a single page at a time is problematic.  A possible workaround is to implement readahead heuristics, where you try to guess what pages sqlite will want to request next and preload them.  It might also be possible to create multiple connections to the same database and share a common cache of pages to avoid redundant reads.  But these workarounds add significant complexity, and while potentially worthwhile when also getting the benefits of sqlite's SQL query execution engine, are not really worthwhile when using sqlite as a plain key-value store.
- for individual values larger than the page size, sqlite stores them as a linked list of overflow pages, which means it is impossible even with a new implementation to read more than one page of the value concurrently.

I don't have much of any experience with parquet (which I believe is what duckdb uses).   Based on my limited understanding I think in principle parquet could work as a "cloud-native" key-value store, but as I don't think that is the normal/intended use case, I am not sure exactly how practical it would be, e.g. how readily existing tools could be used to write the format correctly to enable efficient random access to individual keys, and how many sequential read requests would be needed to read an individual key.

For comparison, with the current mesh format, we have (independent of the number of objects stored):

1 read request for json metadata specifying sharding parameters, but this also provides necessary mesh format metadata
1 read request for shard index + 1 read request for minishard index (may be cached if multiple sequential object ids are requested)
1 read request for manifest
1 read request for each fragment

Thus a total of 5 sequential requests required to download the mesh.

With OCDBT, we have:

1 read request for manifest.ocdbt
1 read request for root b+tree node
n additional read requests for descendant b+tree nodes depending on number of keys and target node size
1 read request for manifest
1 read request for each fragment

Probably in most cases there would be at most 1 additional b+tree node required after the root b+tree node.

**Comment by unidesigner** [comment]
For added context: https://icechunk.io/faq/

**Comment by jbms** [comment]
@unidesigner I'm guessing you referenced that FAQ for the general comments about various storage formats, not specifically to suggest icechunk as a possible container format.  However, mesh data is something that someone might potentially wish to store alongside array data inside an icechunk repository.  However, because icechunk currently only supports data that "looks like" zarr v3, we would have to munge the keys and create fake zarr array metadata in order to actually store it in icechunk.  Additionally, because icechunk has just a single level index of chunk keys (similar to OCDBT with a height of 0), it would not work well if there are a large number of objects.

It seems plausible that the icechunk team may be interested in supporting mesh storage, though, in which case they might extend the format to more readily support storing non-zarr data.

**Comment by unidesigner** [comment]
@jbms Yes, it was meant for general overview of format options, and it includes Tensorstore with ocdbt at the end. I did not want to suggest to use icechunk for mesh storage.

The association of mesh data with array data, as well as other derived data types (skeletons, object features etc.) is a very common usage case and important to think about. Assocating objects in datasets of different type based on an object ID scoped e.g. to a yet-to-defined collection node in the metadata schema would be convenient (i.e. not necessarily required to be a subfolder within a zarr folder hierachy).

Can you explain again what you mean by icechunk with only a single level index not working well for large number of objects? Do you mean that as the number of objects or keys increases, the lookups will take progressively longer without an additional lookup tree, which would be solved using OCDBT?

As an additional remark if the plan is to come up with a better scalable mesh format to also consider efficient storage of additional features on vertices/faces, for static meshes with static and time-dependent features, and perhaps also for time-dependent meshes (where number of vertices and topology may change) and associated features. A use case I once had was coloring synaptic contact areas on neuronal meshes. It would have been nice if this could be supported one day in neuroglancer at scale. And of course the evolution of electric membrane potential across a neuronal mesh in simulations. :)

**Comment by kephale** [comment]
Sorry about the delay, I wanted to share out the RFC update after getting more done on the prototype but probably should have just posted here first. Anywho, I updated the RFC to (hopefully? @normanrz) align with Collections: https://github.com/kephale/ngff/blob/mesh/rfc/8/index.md

**Comment by normanrz** [comment]
I had this tab open for a few weeks and now have some time to respond. Thanks for putting this together and for incorporating our collection ideas. Just as a caveat, the collections proposal is still super early, so it might change a lot over the next couple of months. I'll keep you updated.
I think the proposal needs a few more details about how the mesh files are stored and what their format is. It would be fine to link to the neuroglancer docs and detail any changes, that we might want to make. The "Storage Layout" section says "unsharded"; is that intentional? For datasets with billions of segments, some form of sharding (e.g. the current sharding or ocdbt) will be essential.

**Comment by kephale** [comment]
Awesome, thank you @normanrz. I didn't think this was final and the sharded bit can be cleared up. The more I looked into Collections the more it made sense to build on, but is there something I can do to help nudge the Collections RFC along? Since it is so early as you say, it is hard for me to wrap my head around. 

My goal here was to "rebase" the mesh RFC onto Collections, get some ideas, but then switch back to poking at an actual proof of concept implementation (presumably regardless of the specific RFC most of the code will be reusable). I have something that is close to working that I'll share out the next chance I have to clean it up!

**Comment by jbms** [comment]
I think that it would make sense to specify meshes independently of collections, and have their own standalone metadata in some form, just as we can specify arrays and multiscale arrays independent of collections, but then collections can refer to meshes just as they can refer to other things.

Having the meshes stored somewhere on disk without any metadata would be unfortunate, I think.

However, if you are reluctant to define a standalone metadata format for meshes, we could just use the existing precomputed mesh format exactly, including its `info` JSON file as metadata, and then a collection could just refer to that.  The main drawback, I think, would be the inability to have dimension names other than x, y, z and units other than meters.

**Comment by kephale** [comment]
Howdy Friends,

This thread has been quiet for a while, but I have some progress I'm happy with:

- the RFC is updated based on @jbms' suggestion about not depending on Collections: https://github.com/kephale/ngff/blob/mesh/rfc/8/index.md
- I now have an implementation demo:

  - Generate the demo dataset w/ multiscale meshes: https://github.com/kephale/atrium.kyleharrington.com/blob/main/meshes/complete_blobs_demo/0.0.1.py
  - Inspect in napari: https://github.com/kephale/atrium.kyleharrington.com/blob/main/meshes/view_in_napari/demo.py

Note that I haven't tried to implement multiscale mesh rendering in napari so it just opens all the LODs 🤷 

![Image](https://github.com/user-attachments/assets/1488f667-1dc6-4e79-8439-d0980a2fa16a)

**Comment by joshmoore** [comment]
All I'm hearing is https://forum.image.sc/t/ngff-weekly-dev-update-thread/110810, @kephale! 😄 🎉

**Comment by normanrz** [comment]
Thanks! I think this is shaping up really nicely. I like that you made an implementation to create these meshes. Even that is really valuable.

**Comment by unidesigner** [comment]
I also think that that meshes should be specified independent of collections, and if they are not based on zarr, perhaps get their own type (`OME-Mesh`?). And a collection (`OME-Collection`?) can refer to mesh (OME-Mesh), array (OME-Zarr), ... datasets with their defining feature that all its members (i.e. with their transformations) meaningfully map to a common coordinate frame / (template) space. In my view, it does not make sense to have collections whose members are array or mesh datasets within `OME-Zarr`. It's a bit off-topic and happy to continue the discussion elsewhere when @normanrz shifts his attention to collections.

**Comment by kephale** [comment]
@EricMoerthVis that is the best part. The format basically *is* Neuroglancer precomputed with a `zarr.json` to connect it to the rest of the NGFF dataset, and @jbms is already in the thread!

**Comment by EricMoerthVis** [comment]
That is perfect! Thank you for the clarification!
We (Hidive Lab / former Gehlenborg Lab at Harvard Medical School) are currently working on a NeuroGlancer View integration into Vitessce (https://vitessce.io/) so we are super excited to have scalable mesh and volume formats being supported!

**Comment by jbms** [comment]
Note that the following fields are not supposed to be in the neuroglancer mesh info file:

```
            # Add required fields for viewer compatibility
            "data_type": self.data_type,  # Standard for segmentation data
            "num_channels": 1,      # Single channel for mesh data
            "type": "segmentation",  # Important for Neuroglancer to recognize as meshes
            "scales": [
                {
                    "chunk_sizes": [[64, 64, 64]],
                    "encoding": "compressed_segmentation",
                    "key": "64_64_64",
                    "resolution": [1, 1, 1],
                    "size": [256, 256, 256],  # Match the size of our data
                    "voxel_offset": [0, 0, 0]
                }
            ]
```

Those fields are for a neuroglancer precomputed volume metadata file, not for a mesh metadata file.

**Comment by EricMoerthVis** [comment]
Following up, I tried to load the created Meshes in Neuroglancer and it was unable to read the IDs from the given precomputed format created by your script @kephale. I guess this is related to the point of @jbms.

**Comment by EricMoerthVis** [comment]
@kephale when I create the blobs and use the "view_in_neuroglancer" demo the output looks like this. Is this intended? Looks like the meshes are occupying the same space?

![Image](https://github.com/user-attachments/assets/2a6815b9-c531-4561-9b4a-98f31b985444)

**Comment by kephale** [comment]
Hey @EricMoerthVis, 

It is currently buggy. I use NG less than napari so it was more intuitive to get the napari version working, buuut it does open in NG now 🤣 

My thought was to get the NG version working properly now, then: make corresponding updates to the napari implementation + update the RFC. 

Any input would be greatly appreciated!

**Comment by EricMoerthVis** [comment]
That sounds good!
I will try my best to understand why it doesn't look alright.
I think it has to be somewhere in the code when creating and saving the fragments.

**Comment by fcollman** [comment]
I'm sorry that I didn't chime in here sooner!  First off I understand and support the idea of getting mesh support into the NGFF spec sooner rather than later, and precomputed is definitely is highly functional for static spatial only data.  OME-zarr supports voxel based segmentation but without a mesh spec that is tied to it one has to make an unfortunate sidecar of meshes.  So let's do it. I really don't want anything i'm going to say after this to slow down that train.

I think I might only add a ... yes and... I think we need to also be thinking about what comes next.  

the other issues I see needing addressed long term are:

1) As Jeremy said, it is limited to only spatial + channel data right now.  If someone has a time series of cells they have segmented and want to have a mesh for each time point, they won't be happy with this i don't think. 

1b) More generally, all the same reasons someone wants to have a dataset that comes packaged with meshes they can efficiently access for analysis and visualization, also apply to other vector based objects (points, lines, skeletons, streamlines, tracks etc).  This is why precomputed also has specifications for all of those things as well, but the support for spatial indexing and other access patterns are not uniform.  Meshes can't have vertex properties, but skeletons and  annotations can.  Meshes have a spatial index, but only on a get_by_id type api.  Skeletons don't have a spatial index at all.  Annotations have get_by_group/relationship API, at the expense of data duplication, meshes and skeletons do not. We could just add each of those as seperate elements of a spec, but I do think there are a lot of shared issues and patterns.

2) People are always layering metadata on top of all these objects (at the level of vertex attributes, object attributes, and groups of object attributes).  Often these aren't the same people writing the raw data, and as you start moving into these more derived products there is a need to be able to layer different versions and different layers of metadata on top.  So group 1 writes a giant zarr, group 2 runs a new meshing algorithm on the zarr, group 3 analyzes the meshes and skeletonizes them, group 4 calculates some mesh vertex properties, group 5 skeletonizes the volume, group 6 adds synapses to the volume, etc etc.  

An ideal situation would be that these groups could all reference one another's work and simply have a new zarr.json that references the different data layers, including ones they made and the ones others made.  Neuroglancer states seem to serve this role right now for me, where you have different layers and different data sources that bring together the different elements of the data in your dataset, but it also intertwines all the visualization parameters with the basic definition of those linked data elements.  It feels to me that a vision for NGFF might capture the essence of writing down a series of data elements that are conceptually linked together, without the aspects of visualization.  Precomputed does that at the level of meshes, skeleton and segmentation, but leaves annotations seperated, and doesn't let people layer on more data within a data type (vertex properties on top of meshes (like mesh segmentation). 

3) I personally think one of the key dynamics of OME-zarr success is the zarr component, having a generally accepted and widely implemented base layer of storage access led to many different communities feeling like they had the tools to move forward with their projects that build off this base.  The sharding concept in precomputed I think is really a powerful pattern, and it isn't supported in zarr, but at least to me it feels like it could be at some level of abstraction, it just requires zarr to think more generally about the pattern of how to reshuffle keys into smaller numbers of files.  So pushing on having these key primitives implemented at the zarr level would expand out the impact. 

These were some of the meta things I had in my mind when I was writing up this zarr-vectors spec as a thought experiment of whether there was a way to bring these things together under a more uniform framework with the ambition of having the underlying storage backend backed by zarr.  (https://alleninstitute.github.io/zarr_vectors/)

**Comment by normanrz** [comment]
> 2\. People are always layering metadata on top of all these objects (at the level of vertex attributes, object attributes, and groups of object attributes).  Often these aren't the same people writing the raw data, and as you start moving into these more derived products there is a need to be able to layer different versions and different layers of metadata on top.  So group 1 writes a giant zarr, group 2 runs a new meshing algorithm on the zarr, group 3 analyzes the meshes and skeletonizes them, group 4 calculates some mesh vertex properties, group 5 skeletonizes the volume, group 6 adds synapses to the volume, etc etc.
> 
> 
> An ideal situation would be that these groups could all reference one another's work and simply have a new zarr.json that references the different data layers, including ones they made and the ones others made. Neuroglancer states seem to serve this role right now for me, where you have different layers and different data sources that bring together the different elements of the data in your dataset, but it also intertwines all the visualization parameters with the basic definition of those linked data elements. It feels to me that a vision for NGFF might capture the essence of writing down a series of data elements that are conceptually linked together, without the aspects of visualization. Precomputed does that at the level of meshes, skeleton and segmentation, but leaves annotations seperated, and doesn't let people layer on more data within a data type (vertex properties on top of meshes (like mesh segmentation).

I think this is where [RFC-8 Collections](https://github.com/ome/ngff/pull/343) has an opportunity to tie things together. 



> 3\. The sharding concept in precomputed I think is really a powerful pattern, and it isn't supported in zarr

Fwiw, Zarr v3 has a sharding concept. But it is a different kind of sharding (spatial) than precomputed (hash-based).

I am looking forward to diving deeper into zarr_vectors!

**Comment by fcollman** [comment]
That collections concept does indeed seem very relevant and addresses I think most of the things i brought up!  I love it. 

The one minor exception when it comes to skeletons/points/meshes, is that you might want to add just a vertex attribute array to an existing set of meshes, and if we have a mesh object node (using RFC8 collections nomeclature) a vertex attribute array on that mesh object doesn't really make sense without that mesh object, so it's more directly a kind of sub-node.  Or really i'd like to be able to build a new mesh node, where i reference an existing zarr array virtually, and insert my own zarr array somewhere within the spec for the node type.  This is a slightly different feature I think, than the collections idea as I read it, where you have a complete spec for a node, and now you are just discussing the collections of nodes which relate to some dataset.  

So in my toy example, collections let group 1 write a zarr, group 2 add a mesh object, group 3 to add a skeleton node, group 5 to add their own skeleton node,  group 6 to add a synapse node, but group 4 is stuck with their vertex properties on the meshes, and can't write them into group 2's mesh node, and has to copy the data to meet the mesh node spec, rather than referencing their vertex/link/fragment arrays and writing data that is indexed appropriately to be referenced together.  

Am I making sense? Or is there a way you see the collections spec could address that.. also should i put this comment there? I feel confused only in the sense of OME-zarr 1.0 seems pretty focused on voxel grids, and I think this comment doesn't make sense in that context. 

Re: sharding.. yes i should have been more clear, I understand v3 has sharding in the narrower sense of grouping nearby chunks together, just not the more general key hashing scheme precomputed uses for meshes, annotations, etc.  The spatial sharding is still useful in the zarr-vectors spec to minimize the number of files created but still support spatial chunks.

**Comment by ekatrukha** [comment]
Hello everyone,  my 2 cents are that (as mentioned before) it would be _really_ nice to have meshes with time option from the start. Maybe not in a complicated "time deforming vertices" option, for now, just drop a separate mesh at each timepoint. 
Maybe as an example, something like [Alembic](https://en.wikipedia.org/wiki/Alembic_(computer_graphics)) can be used? Then it is also pretty simple in import to Blender/Maya, etc.
